Finalize write upon receiving SIGINT / SIGTERM

flomnes opened the pull request:
This should ensure that a valid zip archive is produced even if the
user pushes "Ctrl+C".

flomnes committed:
Bump to CXX 17

flomnes committed:
Clarify execution-info.ini file

flomnes committed:
Merge pull request #803 from AntaresSimulatorTeam/fix/execution-info-format

Clarify execution-info.ini file

flomnes committed:
Merge pull request #779 from AntaresSimulatorTeam/feature/cxx-17

Bump to CXX 17 from CXX 11

flomnes committed:
Fix missing renewable columns in districts (sets of areas) (#802)

flomnes committed:
execution_info.ini : add OR-Tools used & solver (#820)

* execution_info.ini : add OR-Tools used & solver

* PR remarks

flomnes committed:
Fix wrong number of cores + code de-duplication (#793)

* Fix wrong number of cores + code de-duplication

* PR remarks + small refactor (template)

flomnes committed:
Release/8.3.1 (#818)

flomnes committed:
Finalize write upon receiving SIGINT / SIGTERM

This should ensure that a valid zip archive is produced even if the
user pushes "Ctrl+C".

flomnes committed:
Avoid double-free

flomnes committed:
Some formatting

flomnes committed:
Refactor + add logs

flomnes committed:
Simplify tuneSolverSpecificOptions (#829)

Use MPSolver::ProblemType

flomnes committed:
[skip ci] Remove useless {}

flomnes committed:
Restore notifications from Sonar (#828)

Source: https://github.com/AntaresSimulatorTeam/antares-xpansion/pull/501

flomnes committed:
Statistics: use std::mutex / std::atomic<> to prevent concurrent writes (#838)

* Statistics: use std::mutex / std::atomic<> to prevent concurrent writes

* Apply PR remarks + renaming

flomnes committed:
Base for BC dual values (#749)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

guilpier-code committed:
doc : new input and output format due to addition of BC marginal prices (#836)

* doc : new input and output format due to addition of BC marginal prices

* BC marginal prices doc : the folder containing the binding constraints output is renamed

guilpier-code committed:
BC marginal price (#801)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

Co-authored-by: Florian OMNES <florian.omnes@rte-france.com>

flomnes committed:
Revert "BC marginal price (#801)"

This reverts commit a6fffd91c0087db4cb3d4532989f3adbb9e44047.

flomnes committed:
Feature/bc marginal price (#845)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

flomnes committed:
BC. MARG. COST: use 2 decimals for the output value (#855)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* BC. MARG. COST: use 2 decimals for the output value

Reading the changelog of version 6.5, I came accros this fix note in
version 6.5

> - Output: the link-variable "MARG.COST" was rounded to an integer
  value (changed to 2 decimal accuracy)

So I thought that we might use 2 decimals for BC. MARG. COST as well.

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

guilpier-code commented on src/solver/application.h:
```diff
@@ -54,6 +54,8 @@ class Application : public Yuni::IEventObserver<Application, Yuni::Policy::Singl
 
     void writeExectutionInfo();
 
+    void installSignalHandlers() const;
```
**installInterruptSignalHandlers()** ?

why **install** ? Rather **initialize** ?

guilpier-code replied:
Why would we have to make  **installSignalHandlers** a method of class **Application** and make this class grow ?

guilpier-code commented on src/solver/main.cpp:
```diff
@@ -111,6 +111,9 @@ int main(int argc, char** argv)
 
     Antares::Solver::Application application;
+
+    application.installSignalHandlers();
```
I guess initializing interruption signals whereas the study and the writer are not created yet is dangerous, as catching such a signal would lead to use them, right ?
Could we not postpone this initialization ?

flomnes replied:
Ok, this method doesn't need to be in class `Application`.

guilpier-code commented on src/solver/signal-handling/common.cpp:
```diff
@@ -0,0 +1,33 @@
+
+static void finalizeWrite()
+{
+    auto study = Data::Study::Current::Get();
```
Please read remarks on **application.h** first (we suppose here that initializing signals handling is not made inside class Application).

I'm not a fan of using the singleton to get the study. It hides the dependency to the study.
What we really need here is the writer (the study only gives the writer).
So I would define a class **SignalHanlder**, create an instance of it in **application::prepare(...)** (after study is created), passing the **study::writer** to its constructor.
**SignalHanlder** could even be a parent class, with children for handling signals on Windows and Linux; we would have a factory like : 
```C++
SignalHanlder signalHandlerFactory(iResultWriterPtr writer)
{
    #ifdef YUNI_OS_WINDOWS
    return winSignalHanlder(writer);
    #else
    return linuxSignalHanlder(writer);
    #endif
}

```
Just an idea, let's discuss it

flomnes replied:
However, we can't pass an `IWriter` object to the functions being called.

You can't provide parameters to signal-handling functions, see this [SO answer](https://stackoverflow.com/a/6970238). Also, signal handlers are global variables.

flomnes replied:
As for the factory method, I disagree since we don't want `winSignalHanlder` compiled on Linux and so on.

guilpier-code replied:
ok, so you don't mind initializing signal handler even if the study is not ready yet. If study is not created when user presses ctrl + c, Antares Simulator stops with an  error log message

guilpier-code commented on src/solver/writer/zip_writer.cpp:
```diff
@@ -70,12 +78,14 @@ ZipWriter::ZipWriter(std::shared_ptr<Yuni::Job::QueueService> qs,
 
 void ZipWriter::addJob(const std::string& entryPath, Yuni::Clob& entryContent)
 {
+    if (pState != ZipState::can_receive_data)
```
We cannot make **addJob** a template method (with **entryContent**'s type as template parameter) because of polymorphism with parent class **IResultWriter**.
But what we could do is have a template private method : 
```C++
class ZipWriter : public IResultWriter
{
....
private:
    template<ContentType>
    void runJob(const std::string& entryPath, Yuni::Clob& entryContent);
}
....

template<class ContentType>
ZipWrite<ContentType>::runJob(const ContentT& entryPath, Yuni::Clob& entryContent)
{
    if (pState != ZipState::can_receive_data)
        return;

    EnsureQueueStartedIfNeeded ensureQueue(this, pQueueService);
    pQueueService->add(
      new ZipWriteJob<ContentType>(*this, entryPath, entryContent, pDurationCollector),
      Yuni::Job::priorityLow);
}
```
and call this template method inside **addJob** methods.
ok, we would have to specialize **runJob** for **IniFile** type...

flomnes replied:
templates don't work well with virtual methods, [see this](https://stackoverflow.com/a/7968213)

guilpier-code replied:
We agree on that, my comment does not contradict yours

flomnes replied:
I've done the requested change [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/827/commits/ae0fa71bf6a01b0e7766d462548cbeb1d2b5f976)

guilpier-code commented on src/solver/writer/zip_writer.cpp:
```diff
@@ -34,6 +35,12 @@ ZipWriteJob<ContentT>::ZipWriteJob(ZipWriter& writer,
 void ZipWriteJob<ContentT>::onExecute()
 {
+    // Don't write data if finalize() has been called
+    if (pState != ZipState::can_receive_data)
```
In case of ctrl + c, if the goal here is let the started jobs finish but prevent others to start, 
maybe the block `if (pState != ZipState::can_receive_data) {...}` is not necessary, as we have such block in the **ZipWriter::addJob(...)** ?

flomnes replied:
We want to prevent new jobs from being added to the queue. This job is already in the queue, `onExecute` is called when the job is, well, executed. If the job is executed after `finalize` has been called, it does nothing.

guilpier-code replied:
ok

flomnes committed:
Fix potential segfault with many BCs, refactor (#853)

* Allocate/deallocate memory to store dual values

* Save address of dual variables for binding constraints

* Use alias variable

* Add binding constraints constants & file names

* bc marginal price : first attempt to add a new output variable in a new direction of the output static algorithm

This commit does not compile, at least on Windows

* BC marginal price : make previous commit compile

* BC marginal price : a bit of enhancement on memory estimating functions in the static output algorithm

* BC marginal price : code saving after storing weekly optimized marginal prices associated to any inequality BC in output variables

* BC marginal price : try to fix build failure on Linux + starting to compute marginal price annual statistics

* BC marginal price : computing marginal price annual statistics

* BC marginal price : avoid too much code duplication when computing marginal price annual statistics

* BC marginal price : set N/A status for a BC output variable depending on BC type and file level

* BC marginal price : printing year by year results

* BC marginal price : handling synthesis results

* BC marginal price : enhancements after review

* BC marginal price : making compilation succeed in Ubuntu

* BC marginal price : making compilation succeed on ubuntu

* BC marginal price : filtering marginal prices results (year-by-year and synthesis) depending on 2 parameters

* BC marginal price : thematic trimming (user can disable output related to BC marginal price)

* BC marginal price : some cleaning + some memory estimation

* BC marginal price : cleaning + restoring + logs

- Cleaning : in the static list of output, call to methods are not required in the context of results related to binding constraint.
- Restoring : oops, previous commit cleaned too much things, a crash occurred
- Logs : the new output variable is now logged just after Areas and Set of areas output variables

* BC marginal price : handling memory estimation for this new variable, in RAM and on disk

* [skip ci] BC marginal price : very small cleaning

* [skip ci] BC marginal price : renaming a function

* BC marginal price : changes after review

* BC marginal price : correction on previous commit

* BC marginal price : fixes so that compilation passes

* BC marginal price : correction due to sonarQube review

* BC marginal price : making compilation pass

* BC marginal price : enhancement due to sonar review

* BC marginal price : changes due to human review

* BC marginal price : making previous commit compile on Linux

* BC marginal price : another try to make compilation pass on Linux

* BC marginal price : small (but important) correction + new try to make compilation pass on Linux

* BC marginal price : replace (pointer + new + delete) with a std::vector + manage wrong initialization

* BC marginal price : (to equal -1, global bc number must be an int) + (rename BC output folder)

if bindConstraintGlobalNumber_ is an unsigned int, it cannot be valued at -1, so we make it an int

* Add missing const qualifier

* BC marginal price : fix a bug in GUI (list of binding constraints not available when it's used)

So we fetch binding constraints where it is available

* Remove commented code

* Remove unused function checkInitialization

* Use switch..case, handle corner cases

* Rename bcInitialize -> isInitialized

* Use switch..case, handle all values

* Apply clang-format

* Use ! instead of not

* Remove duplicate function BuildSurveyReport_noCaptionUpdate

* Apply clang-format

* [skip ci] Remove [[maybe_unused]]

* Fix potential segfault with many BCs, refactor

* Add static_cast<uint>

* Remove lambda return type

* Smell code

* Add TODO note

* PR remarks

* Use std::distance

* Code smells

* More code smells

* Remove unused member

Co-authored-by: Guillaume PIERRE <guillaume.pierre@rte-france.com>

flomnes reviewed (commented):
### Will do
- Take out `initializeSignalHandlers` from class `Application`

### Won't do

- Use templates in a virtual class (impossible)
- Factory method for signal handlers (need too many #ifdef...#endif)
- Avoid using `Study::Get` (impossible). The best I can do is to declare an `IResultWriter::Ptr` as a global variable, and get it from there.

flomnes committed:
Move Application::installSignalHandlers -> free function

flomnes committed:
Add forgotten header file

flomnes committed:
Use template helper function

flomnes commented:
@guilpier-code DONE
Using templatized function for addJob, see ae0fa71b

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/generic-writer-finalize-on-ctrl-c

flomnes committed:
Fix build
